src/Workspaces/Core/Portable/CodeFixesAndRefactorings/DocumentBasedFixAllProviderHelpers.cs

```diff
@@ -30,124 +27,111 @@ internal static class DocumentBasedFixAllProviderHelpers
+            return await ProducerConsumer<(DocumentId documentId, (SyntaxNode? node, SourceText? text))>.RunParallelAsync(
+                source: fixAllContexts,
+                produceItems: static async (fixAllContext, callback, args, cancellationToken) =>
+                {
+                    // Update our progress for each fixAllContext we process.
+                    using var _ = args.progressTracker.ItemCompletedScope();
+
+                    Contract.ThrowIfFalse(
+                        fixAllContext.Scope is FixAllScope.Document or FixAllScope.Project or FixAllScope.ContainingMember or FixAllScope.ContainingType);
+
+                    await args.getFixedDocumentsAsync(fixAllContext, callback).ConfigureAwait(false);
+                },
+                consumeItems: static async (stream, args, cancellationToken) =>
+                {
+                    var currentSolution = args.solution;
+                    using var _ = ArrayBuilder<DocumentId>.GetInstance(out var changedRootDocumentIds);
+
+                    // Next, go and insert those all into the solution so all the docs in this particular project
+                    // point at the new trees (or text).  At this point though, the trees have not been cleaned up.
+                    // We don't cleanup the documents as they are created, or one at a time as we add them, as that
+                    // would cause us to run cleanup on N different solution forks (which would be very expensive).
+                    // Instead, by adding all the changed documents to one solution, and then cleaning *those* we
+                    // only perform cleanup semantics on one forked solution.
+                    await foreach (var (docId, (newRoot, newText)) in stream)
+                    {
+                        // If we produced a new root (as opposed to new text), keep track of that doc-id so that we
+                        // can clean this doc later.
+                        if (newRoot != null)
+                            changedRootDocumentIds.Add(docId);
+
+                        currentSolution = newRoot != null
+                            ? currentSolution.WithDocumentSyntaxRoot(docId, newRoot)
+                            : currentSolution.WithDocumentText(docId, newText!);
+                    }
+
+                    return (currentSolution, changedRootDocumentIds.ToImmutableAndClear());
+                },
+                args: (getFixedDocumentsAsync, progressTracker, solution),
+                cancellationToken).ConfigureAwait(false);
         }
 
+        async Task<Solution> CleanSolutionAsync(Solution dirtySolution, ImmutableArray<DocumentId> changedRootDocumentIds)
         {
+            if (changedRootDocumentIds.IsEmpty)
+                return dirtySolution;
+
+            // Clear out the progress so far.  We're starting a new progress pass for the final cleanup.
+            progressTracker.Report(CodeAnalysisProgress.Clear());
```

Comment:
>progressTracker.Report(CodeAnalysisProgress.Clear()); [](http://example.com/codeflow?start=12&length=53)

Out of curiosity, how is this expressed to the user? Does a bar fill during GetInitialUncleanedSolutionAsync, empty here, and then fill again during this method?

Reply:
Yes. Exactly